### ISSUE TITLE
Call queries from rule left-hand side

### ISSUE THREAD

**Issue by rbrush** [rbrush/clara-rules#172]
Sometimes it's useful to reuse a query in other queries or rules. We can support this by allowing queries to be called in the left-hand side of other productions, and the query results can be unified and filtered just like any other fact.

Consider the following query that finds "cold" locations:

```clj
(defquery is-cold-at
  [:?location]
  [Temperature (< temperature 20) (= location ?location)])
```

We can make it possible to call the query with the parameters and unify against the results from another query. 

```clj
(defquery cold-and-windy-locations
   []
   [WindSpeed (> 30 windspeed) (= ?loc location)]
   [(query is-cold-at :?location ?loc)])
```

So cold-and-windy-locations calls the is-cold-at query with its location to narrow down results. In trivial cases like this it would be just as easy to check the temperature at the location, but this allows reuse of more sophisticated queries.

The query results are just facts, with one match for each result produced. So we could do negations with something like ```[:not [(query is-cold-at :?location ?loc)]]``` or accumulate over query results ```                                      [?all-cold <- (acc/all) :from [(query is-cold-at :?location ?loc)]]```. 

Query results can also be destructured and have additional constraints. Individual query results are just Clojure maps and can be destructured and used as one would expect. For instance, suppose a rule wanted to filter down to locations that start with "A":

```clj
 [(query is-cold-at :?location ?loc) [{location :?location}] (.startsWith location "A")]
```

Basically, anything you can do with a fact you can do with a query result. I think this keeps the mental model simple and intuitive.

**Comment by devn** [comment]
I like this a lot. It makes queries more of a power tool than they currently are, and introduces a sort of named, parameterizable accumulator, no?
